### PR TITLE
Make sample group ID more visible on library (DetailedSample)

### PR DESCRIPTION
**Pull request by callunity**
GLT-2042
* displays the effective (nearest) sample Group ID on both

### CODE REVIEW

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java [hunk 1]
```diff
@@ -159,6 +159,17 @@ public void setGroupDescription(String groupDescription) {
     this.groupDescription = nullifyStringIfBlank(groupDescription);
   }
 
+  @Override
+  public String getNearestSampleGroupId(boolean checkSelf) {
+    DetailedSample start = this;
+    if (!checkSelf) start = this.getParent();
+    for (DetailedSample sample = start; sample != null; sample = sample.getParent()) {
+      String groupId = sample.getGroupId();
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java]: MAJOR Rename "groupId" which hides the field declared at line 58. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)

**Comment by sqwci** [comment]
SonarQube analysis reported 52 issues
* ![CRITICAL][CRITICAL] 13 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 35 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DetailedSampleBuilder.java#L981](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleBuilder.java#L981): Define a constant instead of duplicating this literal "Cannot delete non-persisted builder object" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L546](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L546): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L572](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L572): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L609](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L609): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L810](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L810): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L828](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L828): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L851](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L851): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L913](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L913): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L928](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L928): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L1001](https://github.com/TGAC/miso-lims/blob/2107712c1a2fc430633344a6af8f794e1489e1cc/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L1001): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java [hunk 2]
```diff
@@ -159,6 +160,15 @@ public void setGroupDescription(String groupDescription) {
     this.groupDescription = nullifyStringIfBlank(groupDescription);
   }
 
+  @Override
+  public Optional<DetailedSample> getEffectiveGroupIdSample() {
+    for (DetailedSample sample = this; sample != null; sample = sample.getParent()) {
+      String groupId = sample.getGroupId();
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleImpl.java]: MAJOR Rename "groupId" which hides the field declared at line 59. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)

**Comment by sqwci** [comment]
SonarQube analysis reported 52 issues
* ![CRITICAL][CRITICAL] 13 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 35 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DetailedSampleBuilder.java#L982](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleBuilder.java#L982): Define a constant instead of duplicating this literal "Cannot delete non-persisted builder object" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L548](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L548): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L574](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L574): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L625](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L625): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L830](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L830): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L848](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L848): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L871](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L871): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L933](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L933): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L948](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L948): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L1021](https://github.com/TGAC/miso-lims/blob/d5e3237843936663707c10fbe36c958930a2cd47/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L1021): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java [hunk 3]
```diff
@@ -356,6 +357,11 @@ private static DetailedSampleDto asDetailedSampleDto(DetailedSample from) {
       dto.setParentTissueSampleClassId(from.getParent().getSampleClass().getId());
       dto.setIdentityConsentLevel(getIdentityConsentLevelString(from));
     }
+    Optional<DetailedSample> effective = from.getEffectiveGroupIdSample();
+    if (effective.isPresent()) {
```
Comment (apmasell) [inline miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java]: You can do:

```
effective.ifPresent(parent -> {
dto.setEffectiveGroupId(parent.getGroupId());
dto.setEffectiveGroupIdSample(parent.getAlias());
});
```

File: miso-web/src/main/webapp/pages/editLibrary.jsp [hunk 4]
```diff
@@ -374,6 +374,15 @@
 <br/>
 <h2>Details</h2>
 <table class="in">
+  <tr>
+    <td class="h">Upstream Group ID:</td>
+    <td id="effectiveGroupId">
+      <c:choose>
+      <c:when test="${empty effectiveGroupId}">(None)</c:when>
+      <c:otherwise>${effectiveGroupId}</c:otherwise>
```
Comment (djcooke) [inline miso-web/src/main/webapp/pages/editLibrary.jsp]: Didn't you also want to show the alias of the entity that this is set on? Same on Edit Sample page and the bulk pages

Comment (callunity) [reply miso-web/src/main/webapp/pages/editLibrary.jsp]: Yes! But only on the Edit pages; the bulk pages already have enough columns that I'm going to leave the alias off for now, and we can add it later if it's desired.

Comment (djcooke) [reply miso-web/src/main/webapp/pages/editLibrary.jsp]: :+1:

**Comment by sqwci** [comment]
SonarQube analysis reported 52 issues
* ![CRITICAL][CRITICAL] 13 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 35 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DetailedSampleBuilder.java#L982](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/DetailedSampleBuilder.java#L982): Define a constant instead of duplicating this literal "Cannot delete non-persisted builder object" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L548](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L548): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L574](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L574): Define a constant instead of duplicating this literal "library" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L625](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L625): Define a constant instead of duplicating this literal "Libraries" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L830](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L830): Define a constant instead of duplicating this literal "Dilutions" 4 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L848](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L848): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L871](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L871): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L933](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L933): Define a constant instead of duplicating this literal "_POOL" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L948](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L948): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)
1. ![CRITICAL][CRITICAL] [EditLibraryController.java#L1021](https://github.com/TGAC/miso-lims/blob/347ef5b73458c4dd312fe1d8598ffd4936b104da/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditLibraryController.java#L1021): Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**